packages/hardhat/contracts/strategy/CrabMigration.sol

```diff
@@ -99,44 +125,117 @@ import {StrategyMath} from "./base/StrategyMath.sol";
+
+        // 2. Callback
+        _flashCallback(data.caller, data.amountToBorrow, data.callSource, data.callData);
+
+        // 3. Repay the weth:
+        weth.deposit{value: data.amountToBorrow}();
+        weth.approve(EULER_MAINNET, type(uint256).max);
+        IEulerDToken(dToken).repay(0, data.amountToBorrow);
+    }
+
+    function _flashCallback(
+        address _initiator,
+        uint256 _amount,
+        uint8 _callSource,
+        bytes memory _calldata
+    ) internal {
+        if (FLASH_SOURCE(_callSource) == FLASH_SOURCE.BATCH_MIGRATE) {
+            // deferLiquidityCheck can be called by anyone
+            require(_initiator == owner, "M1");
+
+            uint256 crabV1Balance = crabV1.balanceOf(address(this));
 
-        IEulerDToken(dToken).borrow(0, amountEthToBorrow);
-        weth.withdraw(amountEthToBorrow);
+            // 2. mint osqth in crab v2
+            uint256 wSqueethToMint = crabV1.getWsqueethFromCrabAmount(crabV1Balance);
+            crabV2.initialize{value: _amount}(wSqueethToMint);
 
-        // 2. mint osqth in crab v2 
-        uint256 wSqueethToMint = crabV1.getWsqueethFromCrabAmount(crabV1Balance);
-        crabV2.initialize{value: amountEthToBorrow}(wSqueethToMint);
+            // 3. call withdraw from crab v1
+            IERC20(wPowerPerp).approve(address(crabV1), type(uint256).max);
+            crabV1.approve(address(crabV1), crabV1Balance);
+            crabV1.withdraw(crabV1Balance);
+        } else if (FLASH_SOURCE(_callSource) == FLASH_SOURCE.FLASH_MIGRATE_V1_TO_V2) {
+            FlashMigrateV1toV2 memory data = abi.decode(_calldata, (FlashMigrateV1toV2));
+            crabV2.deposit{value: _amount}();
+            uint256 crabV1ToWithdraw = crabV1.balanceOf(_initiator);
 
-        // 3. call withdraw from crab v1
-        IERC20(wPowerPerp).approve(address(crabV1), type(uint).max);
-        crabV1.approve(address(crabV1), crabV1Balance);
-        crabV1.withdraw(crabV1Balance);
+            crabV1.transferFrom(_initiator, address(this), crabV1ToWithdraw);
+            IERC20(wPowerPerp).approve(address(crabV1), data.v1oSqthToPay);
+            crabV1.withdraw(crabV1ToWithdraw);
 
-        // 4. Repay the weth:
-        weth.deposit{value: amountEthToBorrow}();
-        weth.approve(EULER_MAINNET, type(uint).max);
-        IEulerDToken(dToken).repay(0, amountEthToBorrow);
+            // Flash deposit remaining ETH, if user said so. Else return back the ETH. If CR1 = CR2 ethToFlashDeposit should be 0
+            if (data.ethToFlashDeposit > 0) {
+                crabV2.flashDeposit{value: address(this).balance - _amount}(data.ethToFlashDeposit);
+            }
+
+            // Sent back the V2 tokens to the user
+            crabV2.transfer(_initiator, crabV2.balanceOf(address(this)));
+
+            // Sent back the excess ETH
+            if (address(this).balance > _amount) {
+                payable(_initiator).sendValue(address(this).balance - _amount);
```

Comment:
the good practice would be for us to keep track of this value and not just send the difference, since if someone sent additional eth to the contract etc it sends more eth to the user. Not the biggest issue for rn though